spec/models/spree/gateway/amazon_spec.rb

```diff
@@ -38,16 +51,148 @@
+
+        expect(response).to be_success
+      end
+    end
+
+    context 'when declined' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/xml'},
+          status: 200,
+          body: build_mws_auth_declined_response(order: order),
+        })
+
+        response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+
+        expect(response).not_to be_success
+        expect(response.message).to eq('Authorization failure: InvalidPaymentMethod')
+      end
+    end
+
+    context 'with an ErrorResponse error' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/xml'},
+          status: 400,
+          body: build_mws_auth_error_response(order: order),
+        })
+
+        response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+
+        expect(response).not_to be_success
+        expect(response.message).to match(/^400 TransactionAmountExceeded:/)
+      end
+    end
+
+    context 'with a 5xx error' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/plain'},
+          status: 502,
+          body: 'Bad Gateway',
+        })
+
+        response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+
+        expect(response).not_to be_success
+        expect(response.message).to match(/502 Bad Gateway/)
+      end
+    end
+
+    # 500 is special-cased in the Amazon library
+    context 'with a 500 error' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/plain'},
+          status: 500,
+          body: 'Server Error',
+        })
+
+        expect {
+          response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+        }.to raise_error(Spree::Core::GatewayError, 'InternalServerError')
+      end
+    end
+
+    # 503 is special-cased in the Amazon library
+    context 'with a 503 error' do
+      before do
+        # Without this the specs have a big pause while the Amazon gem retries
+        expect_any_instance_of(PayWithAmazon::Request).to receive(:get_seconds_for_try_count).at_least(:once).and_return(0)
+      end
+
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/plain'},
+          status: 503,
+          body: 'Service Unavailable',
+        })
+
+        expect {
+          response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+        }.to raise_error(Spree::Core::GatewayError, 'ServiceUnavailable or RequestThrottled')
+      end
+    end
+
+    describe 'sandbox simulation strings' do
```

Comment:
I'd rather use `describe` only for public methods on the class, e.g. `describe "#authorize"` should we move this block as a `context` inside the `describe` block for `#authorize`?

```
describe '#authorize' do
  context "with sandbox simulation strings" do
    ...
  end
end